File: Data_Structure/Linked_list/SinglyLL.py
class Node:
    def __init__(self, data=0, next_node=None):
        self.data = data
        self.next = next_node

class LinkedList:
    def __init__(self):
        self.start = None
        self.size = 0

    def isEmpty(self):
        return self.start is None

    def getListSize(self):
        return self.size

    def viewList(self):
        if self.isEmpty():
            print("Empty List")
            return
        current = self.start
        while current:
            print(f"{current.data} |___|-->", end="")
            current = current.next
        print("null")

    def insertAtFirst(self, val):
        newNode = Node(val, self.start)
        self.start = newNode
        self.size += 1

    def insertAtLast(self, val):
        newNode = Node(val)
        if self.isEmpty():
            self.start = newNode
        else:
            current = self.start
            while current.next:
                current = current.next
            current.next = newNode
        self.size += 1

    def insertAtPos(self, val, pos):
        if pos < 1 or pos > self.size + 1:
            print("Insertion is not possible")
            return
        if pos == 1:
            self.insertAtFirst(val)
            return
        newNode = Node(val)
        current = self.start
        for _ in range(1, pos-1):
            current = current.next
        newNode.next = current.next
        current.next = newNode
        self.size += 1

    def updateData(self, val, pos):
        if self.isEmpty() or pos < 1 or pos > self.size:
            print("Updation is not possible")
            return
        current = self.start
        for _ in range(1, pos):
            current = current.next
        current.data = val

    def deleteFirst(self):
        if self.isEmpty():
            print("List is empty")
            return
        self.start = self.start.next
        self.size -= 1

    def deleteLast(self):
        if self.isEmpty():
            print("List is empty")
            return
        if self.start.next is None:
            self.start = None
        else:
            current = self.start
            while current.next.next:
                current = current.next
            current.next = None
        self.size -= 1

    def deleteAtPos(self, pos):
        if self.isEmpty() or pos < 1 or pos > self.size:
            print("Deletion not possible")
            return
        if pos == 1:
            self.deleteFirst()
            return
        current = self.start
        for _ in range(1, pos-1):
            current = current.next
        current.next = current.next.next
        self.size -= 1

    def reverseList(self):
        previous = None
        current = self.start
        while current:
            next_node = current.next
            current.next = previous
            previous = current
            current = next_node
        self.start = previous


def main():
    linked_list = LinkedList()
    while True:
        print("\n1. Add item to the list at start")
        print("2. Add item to the list at last")
        print("3. Add item to the list at position")
        print("4. Delete first node")
        print("5. Delete last node")
        print("6. Delete node at position")
        print("7. Update node at position")
        print("8. Reverse link list")
        print("9. View list")
        print("10. Get List size")
        print("11. Exit")

        choice = int(input("\nEnter your choice: "))

        if choice == 1:
            value = int(input("Enter value: "))
            linked_list.insertAtFirst(value)
        elif choice == 2:
            value = int(input("Enter value: "))
            linked_list.insertAtLast(value)
        elif choice == 3:
            value = int(input("Enter value: "))
            position = int(input("Enter position: "))
            linked_list.insertAtPos(value, position)
        elif choice == 4:
            linked_list.deleteFirst()
        elif choice == 5:
            linked_list.deleteLast()
        elif choice == 6:
            position = int(input("Enter position: "))
            linked_list.deleteAtPos(position)
        elif choice == 7:
            value = int(input("Enter value: "))
            position = int(input("Enter position: "))
            linked_list.updateData(value, position)
        elif choice == 8:
            linked_list.reverseList()
        elif choice == 9:
            linked_list.viewList()
        elif choice == 10:
            print(f"List size: {linked_list.getListSize()}")
        elif choice == 11:
            break
        else:
            print("Invalid choice.")

if __name__ == "__main__":
    main()

File: Data_Structure/Linked_list/SinglyLL.java
import java.util.Scanner;

public class SinglyLL {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        LinkedLists list = new LinkedLists(); // Instantiate your LinkedLists class

        // Menu to manage the singly linked list
        while (true) {
            System.out.println("\nSingly Linked List Operations Menu:");
            System.out.println("1. Insert at First");
            System.out.println("2. Insert at Last");
            System.out.println("3. Insert at Position");
            System.out.println("4. View List");
            System.out.println("5. Exit");
            System.out.print("Enter your choice (1-5): ");

            int choice = sc.nextInt(); // Read user's choice

            switch (choice) {
                case 1:
                    System.out.print("Enter value to insert at first: ");
                    int value1 = sc.nextInt();
                    list.insertAtFirst(value1);
                    break;
                case 2:
                    System.out.print("Enter value to insert at last: ");
                    int value2 = sc.nextInt();
                    list.insertAtLast(value2);
                    break;
                case 3:
                    System.out.print("Enter value to insert: ");
                    int value3 = sc.nextInt();
                    System.out.print("Enter position to insert: ");
                    int position = sc.nextInt();
                    list.insertAtPos(value3, position);
                    break;
                case 4:
                    System.out.println("Current List:");
                    list.viewList();
                    break;
                case 5:
                    System.out.println("Exiting...");
                    sc.close(); // Close the scanner
                    return; // Exit the program
                default:
                    System.out.println("Invalid choice. Please enter a number between 1 and 5.");
            }
        }
    }
}

class LinkedLists {
    // Node class for singly linked list
    class Node {
        int data; // Store the integer data
        Node next; // Pointer to the next node in the list
    }
    private Node head; // Head points to the first node of the list
    private int size; // Tracks the number of elements in the list

    // Constructor initializes an empty list
    public LinkedLists() {
        head = null;
        size = 0;
    }

    // Returns true if the list is empty
    public boolean isEmpty() {
        return head == null;
    }

    // Inserts a node at the beginning of the list
    public void insertAtFirst(int val) {
        Node n = new Node(); // Create a new node
        n.data = val; // Set the data of the node
        n.next = head; // New node's next pointer points to the current head
        head = n; // Update the head to the new node
        size++; // Increment the size of the list
    }

    // Inserts a node at the end of the list
    public void insertAtLast(int val) {
        Node n = new Node(); // Create a new node
        n.data = val; // Set the data of the node
        if (isEmpty()) {
            head = n; // If the list was empty, set head to the new node
        } else {
            Node t = head;
            while (t.next != null) { // Traverse to the end of the list
                t = t.next;
            }
            t.next = n; // Link the last node's next to new node
        }
        size++; // Increment the size of the list
    }

    // Prints all the elements in the list from head to end
    public void viewList() {
        if (isEmpty()) {
            System.out.println("Empty list"); // Check if the list is empty
            return;
        }
        Node t = head;
        while (t != null) {
            System.out.print(t.data + " -> "); // Print each node's data
            t = t.next; // Move to the next node
        }
        System.out.println("null");
    }

    public void insertAtPos(int val, int pos) {
        if (pos < 1 || pos > size + 1) {
            System.out.println("Invalid position");
            return;
        }
        if (pos == 1) {
            insertAtFirst(val);
            return;
        }
        Node n = new Node();
        n.data = val;
        Node current = head;
        for (int i = 1; i < pos - 1; i++) { // Traverse to the position before where new node will be inserted
            current = current.next;
        }
        n.next = current.next; // Set new node
    }
